### ISSUE TITLE
Bug in JavaDateFormatter when using DOY

### ISSUE THREAD

**Issue by HayDegha0917** [elastic/elasticsearch#89096]
### Elasticsearch Version

7.10.2

### Java Version

1.8

### OS Version

CentOS

### Problem Description

When defining a mapping that has a field that uses DOY format, data gets indexed correctly However, we run into issues when using the rounding parser. See steps below for a test case.

We believe the issue is related to this section of code:

https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/common/time/JavaDateFormatter.java#L48-L59

### Steps to Reproduce

```
DELETE /test-doy-date

PUT /test-doy-date
{
  "mappings": {
    "properties": {
      "timestamp": {
        "type": "date",
        "format": "yyyy-DDD'T'HH:mm:ss.SSS||yyyy-DDD'T'HH:mm:ss.SSSX||yyyy-DDD'T'HH:mm:ss.n||yyyy-DDD'T'HH:mm:ss.nX||strict_date_optional_time||epoch_millis"
      }
    }
  }
}

POST _bulk
{ "index" : { "_index" : "test-doy-date", "_id" : "1" } }
{ "timestamp" : "2022-104T14:08:30.100" }
{ "index" : { "_index" : "test-doy-date", "_id" : "2" } }
{ "timestamp" : "2022-104T14:08:30.540Z" }
{ "index" : { "_index" : "test-doy-date", "_id" : "3" } }
{ "timestamp" : "2022-104T14:08:31.100111" }
{ "index" : { "_index" : "test-doy-date", "_id" : "4" } }
{ "timestamp" : "2022-104T14:08:31.234567Z" }

GET /test-doy-date/_search

GET /test-doy-date/_search
{
  "query": {
    "bool": {
      "filter": {
        "range": {
          "timestamp": {
            "gte": "2022-104T14:08:30.293",
            "lte": "2022-104T14:08:31.355",
            "format": "yyyy-DDD'T'HH:mm:ss.SSS"
          }
        }
      }
    }
  }
}
```

All steps complete without error except for the range filter, which outputs:

```
{
  "error" : {
    "root_cause" : [
      {
        "type" : "parse_exception",
        "reason" : "failed to parse date field [2022-104T14:08:31.355] with format [yyyy-DDD'T'HH:mm:ss.SSS]: [failed to parse date field [2022-104T14:08:31.355] with format [yyyy-DDD'T'HH:mm:ss.SSS]]"
      }
    ],
    "type" : "search_phase_execution_exception",
    "reason" : "all shards failed",
    "phase" : "query",
    "grouped" : true,
    "failed_shards" : [
      {
        "shard" : 0,
        "index" : "test-doy-date",
        "node" : "8QwQl8a5SvWteKZqObjW5g",
        "reason" : {
          "type" : "parse_exception",
          "reason" : "failed to parse date field [2022-104T14:08:31.355] with format [yyyy-DDD'T'HH:mm:ss.SSS]: [failed to parse date field [2022-104T14:08:31.355] with format [yyyy-DDD'T'HH:mm:ss.SSS]]",
          "caused_by" : {
            "type" : "illegal_argument_exception",
            "reason" : "failed to parse date field [2022-104T14:08:31.355] with format [yyyy-DDD'T'HH:mm:ss.SSS]",
            "caused_by" : {
              "type" : "date_time_parse_exception",
              "reason" : "date_time_parse_exception: Text '2022-104T14:08:31.355' could not be parsed: Conflict found: Field DayOfYear 1 differs from DayOfYear 104 derived from 2022-01-01",
              "caused_by" : {
                "type" : "date_time_exception",
                "reason" : "date_time_exception: Conflict found: Field DayOfYear 1 differs from DayOfYear 104 derived from 2022-01-01"
              }
            }
          }
        }
      }
    ]
  },
  "status" : 400
}
```

If we omit the explicit format in the filter, we instead get

```
{
  "error" : {
    "root_cause" : [
      {
        "type" : "parse_exception",
        "reason" : "failed to parse date field [2022-104T14:08:31.355] with format [yyyy-DDD'T'HH:mm:ss.SSS||yyyy-DDD'T'HH:mm:ss.SSSX||yyyy-DDD'T'HH:mm:ss.n||yyyy-DDD'T'HH:mm:ss.nX||strict_date_optional_time||epoch_millis]: [failed to parse date field [2022-104T14:08:31.355] with format [yyyy-DDD'T'HH:mm:ss.SSS||yyyy-DDD'T'HH:mm:ss.SSSX||yyyy-DDD'T'HH:mm:ss.n||yyyy-DDD'T'HH:mm:ss.nX||strict_date_optional_time||epoch_millis]]"
      }
    ],
    "type" : "search_phase_execution_exception",
    "reason" : "all shards failed",
    "phase" : "query",
    "grouped" : true,
    "failed_shards" : [
      {
        "shard" : 0,
        "index" : "test-doy-date",
        "node" : "MB7SIp_pSrSComH5XlvIPQ",
        "reason" : {
          "type" : "parse_exception",
          "reason" : "failed to parse date field [2022-104T14:08:31.355] with format [yyyy-DDD'T'HH:mm:ss.SSS||yyyy-DDD'T'HH:mm:ss.SSSX||yyyy-DDD'T'HH:mm:ss.n||yyyy-DDD'T'HH:mm:ss.nX||strict_date_optional_time||epoch_millis]: [failed to parse date field [2022-104T14:08:31.355] with format [yyyy-DDD'T'HH:mm:ss.SSS||yyyy-DDD'T'HH:mm:ss.SSSX||yyyy-DDD'T'HH:mm:ss.n||yyyy-DDD'T'HH:mm:ss.nX||strict_date_optional_time||epoch_millis]]",
          "caused_by" : {
            "type" : "illegal_argument_exception",
            "reason" : "failed to parse date field [2022-104T14:08:31.355] with format [yyyy-DDD'T'HH:mm:ss.SSS||yyyy-DDD'T'HH:mm:ss.SSSX||yyyy-DDD'T'HH:mm:ss.n||yyyy-DDD'T'HH:mm:ss.nX||strict_date_optional_time||epoch_millis]",
            "caused_by" : {
              "type" : "date_time_parse_exception",
              "reason" : "Failed to parse with all enclosed parsers"
            }
          }
        }
      }
    ]
  },
  "status" : 400
}
```

If we change `lte` to `lt` in the filter, we get a valid response:

```
{
  "took" : 9,
  "timed_out" : false,
  "_shards" : {
    "total" : 5,
    "successful" : 5,
    "skipped" : 0,
    "failed" : 0
  },
  "hits" : {
    "total" : {
      "value" : 3,
      "relation" : "eq"
    },
    "max_score" : 0.0,
    "hits" : [
      {
        "_index" : "test-doy-date",
        "_type" : "_doc",
        "_id" : "3",
        "_score" : 0.0,
        "_source" : {
          "timestamp" : "2022-104T14:08:31.100111"
        }
      },
      {
        "_index" : "test-doy-date",
        "_type" : "_doc",
        "_id" : "4",
        "_score" : 0.0,
        "_source" : {
          "timestamp" : "2022-104T14:08:31.234567Z"
        }
      },
      {
        "_index" : "test-doy-date",
        "_type" : "_doc",
        "_id" : "2",
        "_score" : 0.0,
        "_source" : {
          "timestamp" : "2022-104T14:08:30.540Z"
        }
      }
    ]
  }
}
```

If we use a custom format in the field mapping that does **not** use DOY, everything works fine.

**Comment by nik9000** [comment]
I'll debug this a bit and figure out where it lands.

**Comment by nik9000** [comment]
yeah - it looks indeed like it's https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/common/time/JavaDateFormatter.java#L48-L59. All those default looks wrong if you are using month of year. The reason `lt` works is because it turns off the "rounding" that you bumped into.

FWIW, if I comment out those two bits of rounding it doesn't crash. Not sure if that's the fix, but it doesn't crash. 

Here's the test I'm running locally:
```
curl -uelastic:password -HContent-Type:application/json -XDELETE localhost:9200/test-doy-date

curl -uelastic:password -HContent-Type:application/json -XPUT localhost:9200/test-doy-date?pretty -d'
{
  "mappings": {
    "properties": {
      "timestamp": {
        "type": "date",
        "format": "yyyy-DDD'"'"'T'"'"'HH:mm:ss.SSS||yyyy-DDD'"'"'T'"'"'HH:mm:ss.SSSX||yyyy-DDD'"'"'T'"'"'HH:mm:ss.n||yyyy-DDD'"'"'T'"'"'HH:mm:ss.nX||strict_date_optional_time||epoch_millis"
      }
    }
  }
}
'

curl -uelastic:password -HContent-Type:application/json -XPOST 'localhost:9200/_bulk?pretty&refresh' -d'
{ "index" : { "_index" : "test-doy-date", "_id" : "1" } }
{ "timestamp" : "2022-104T14:08:30.100" }
{ "index" : { "_index" : "test-doy-date", "_id" : "2" } }
{ "timestamp" : "2022-104T14:08:30.540Z" }
{ "index" : { "_index" : "test-doy-date", "_id" : "3" } }
{ "timestamp" : "2022-104T14:08:31.100111" }
{ "index" : { "_index" : "test-doy-date", "_id" : "4" } }
{ "timestamp" : "2022-104T14:08:31.234567Z" }
'

curl -uelastic:password -HContent-Type:application/json -XGET localhost:9200/test-doy-date/_search?pretty

curl -uelastic:password -HContent-Type:application/json -XGET 'localhost:9200/test-doy-date/_search?pretty&error_trace' -d'
{
  "query": {
    "bool": {
      "filter": {
        "range": {
          "timestamp": {
            "gte": "2022-104T14:08:30.293",
            "lte": "2022-104T14:08:31.355",
            "format": "yyyy-DDD'"'"'T'"'"'HH:mm:ss.SSS"
          }
        }
      }
    }
  }
}'
```

**Comment by nik9000** [comment]
I've sent this to the core-infra folks who know this code much better than I do. This is specifically about the "round up" parser.

**Comment by nik9000** [comment]
> Also, please note that this issue does not occur with ES6, presumably because it uses Joda time rather than Java time.

Yeah. I imagine that's what it was. That was a big change and lots of stuff snuck in, unfortunately.



> @nik9000 do you know the conditions that cause ES to use the "round up" parser? It would help us communicate the extent of the issue to our users.

It looks like `gt` and `lte` on `range` queries on `date`, `date_nanos`, and `date_range` fields. That *looks* like it. There are some other places that "round up" but they do it to non-dates. I think. It's not 100% clear everywhere without a bunch of digging.

**Comment by nik9000** [comment]
Sorry about that! Day of year queries aren't something I'd use, but I'm not
in your industry. Anyway!

The tags I've put on this should cause an expert in this stuff to have a
look at it before long. Sometime this week would be normal. But it might
take longer, depending. Summer in the northern hemisphere has many folks
away from their keyboard. Even then, I'm not sure where it's fall on their
schedule. I hate breaking things, but given how long it's been like this
that implies a slightly smaller blast radius.

You can certainly have a look yourself. Like I said, I commented out some
code and everything seemed better. But I didn't run the test suite. I
really may have broken other things. Deleting code can do that....

It'd need to be reviewed by one of those experts. They'd know exactly how
good our tests are here.

After that there's actually cutting the release. We have a policy of never
telling folks when we will cut releases. Mostly it's because the dates
shift a fair bit and we don't want to promise.

**Comment by pgomulka** [comment]
it is the same family of bugs as https://github.com/elastic/elasticsearch/issues/58986

it is because when parsing `yyyy-DDD'T'HH:mm:ss.SSS` we will try to default missing fields like Month etc https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/common/time/JavaDateFormatter.java#L48-L59
So when parsing `2022-104T14:08:30.293` this results in interim
`{SecondOfMinute=30, DayOfMonth=1, YearOfEra=2022, DayOfYear=104, MinuteOfHour=8, MonthOfYear=1, HourOfDay=14, NanoOfSecond=293000000},null`
Note that ther is `MonthOfYear` and `DayOfMonth` defaulted. When java resolves a date part to `2022-01-01`.  See that a monthofyear is used before dayOfYear:
https://github.com/openjdk/jdk/blob/master/src/java.base/share/classes/java/time/chrono/AbstractChronology.java#L439

I think the conclusion is the same as in the linked issue. https://github.com/elastic/elasticsearch/issues/58986#issuecomment-675893674 
There is no way to fix this at the moment.

**Comment by pgomulka** [comment]
@HayDegha0917 I did open an issue to the JDK regarding this https://bugs.java.com/bugdatabase/view_bug.do?bug_id=JDK-8250514 but the problem comes down to ES not being able to know up front what format (and fields effectively) will be used. If we knew what format will be used, we could use the advice from the JDK issue.

The reason this worked previously in 6.x was because joda had [parseInto](https://www.joda.org/joda-time/apidocs/org/joda/time/format/DateTimeParser.html#parseInto-org.joda.time.format.DateTimeParserBucket-java.lang.String-int-) method, which from what I understand was a "predecessor" of parseDefaulting in java.time. Turns out it works differently in some cases.

I am open for ideas and PRs. Do read the [original issue ](https://github.com/elastic/elasticsearch/issues/58986)
